### DISCUSSION TITLE
visit_string_literal

### DISCUSSION THREAD

**Discussion by SKT1yang** [oxc-project/oxc#4820] [Q&A]
Is there a way to change the literal node to a function calling node when accessing a string literal using visit_string_iteral, with the original string literal as the parameter?
<img width="1053" alt="image" src="https://github.com/user-attachments/assets/4b56c2aa-a320-476d-96c2-9e2f7693d2aa">
The error message is:
<img width="1189" alt="image" src="https://github.com/user-attachments/assets/504b8268-774d-474d-a08f-27891796ca5b">

**Comment by overlookmotel** [comment]
A couple of things:

1. In `visit_string_literal` you can only replace the node with another `StringLiteral`. If you want to alter an `Expression` from one type of `Expression` to another, you need to do this in `visit_expression`.
2. It looks like you have `arguments` and `callee` in reverse order. `CallExpression`'s fields are in an unintuitive order (our bad - https://github.com/oxc-project/oxc/issues/4821).

Hope this helps.

**Reply by SKT1yang** [reply]
Thank you for your reply. It has successfully taken effect, but there is another issue. I converted the string literal into a function call with the parameter being the string literal, and then it actually recursively passed!!! The recursive order is: add t function call ->add string literal ->modify to add t function call again ->add string literal again. I couldn't find any useful information in the parameters, such as whether the previous node was a function call, so I couldn't accurately interrupt the recursion
<img width="962" alt="image" src="https://github.com/user-attachments/assets/2a0b9d5c-a171-4d77-a0b0-84d4a67831a6">

**Reply by overlookmotel** [reply]
I'm afraid I can't quite follow what you're saying. But one note: `string_literal.value` is an `Atom`. `Atom` is a wrapper around a `&str`. So I don't think you need to convert to a `String` - just:

1. Pass `string_literal.value.as_str()` to `contains_chinese()` (assuming `contains_chinese` takes a `&str`).
2. Pass `string_literal.value.clone()` to `ast_builder.expression_string_literal()`.

NB: `Atom`s are very cheap to clone.

I don't know if that will solve your problem or not, but it'd be an improvement. If problem continues, please can you try explaining again what's happening? What is logged? Is it an infinite recursion loop?
